File: tests/e2e/nodejs_client/test/export_data.test.js
/**
 * E2E: EXPORT DATA → Cloud Storage (CSV) against a live bqemulator
 * container via the @google-cloud/bigquery Node.js client.
 *
 * EXPORT DATA runs as a query job: the inner SELECT is materialised and
 * written to the wildcard `uri` under `BQEMU_GCS_LOCAL_ROOT`. With a
 * single output shard the `*` expands to a 12-digit zero-padded counter,
 * so `export_nodejs/*.csv` becomes `export_nodejs/000000000000.csv`. The
 * Makefile target (`test-e2e-nodejs`) bind-mounts a host directory onto
 * `/var/lib/bqemu-gcs` and exports its host path as
 * `BQEMU_GCS_HOST_ROOT`, so we read the exported file straight off the
 * mount.
 */

const { describe, it, before, after } = require("node:test");
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");

const REST_URL = process.env.BQEMU_REST_URL || "http://localhost:9050";
const GCS_HOST_ROOT = process.env.BQEMU_GCS_HOST_ROOT;
const PROJECT = "e2e-nodejs-export";
const DATASET = "export_node_ds";
const BUCKET = "g1-e2e";

function makeClient() {
  const { BigQuery } = require("@google-cloud/bigquery");
  const { OAuth2Client } = require("google-auth-library");

  const fake = new OAuth2Client();
  fake.credentials = { access_token: "anonymous" };

  return new BigQuery({
    projectId: PROJECT,
    apiEndpoint: REST_URL,
    authClient: fake,
    autoRetry: false,
  });
}

describe("bqemulator EXPORT DATA → CSV (Node.js)", () => {
  let client;
  before(async () => {
    client = makeClient();
    // Reset any dataset left by a crashed prior run so reruns are deterministic.
    try {
      await client.dataset(DATASET).delete({ force: true });
    } catch (_) {
      /* ignore */
    }
    await client.createDataset(DATASET, { location: "US" });
  });
  after(async () => {
    try {
      await client.dataset(DATASET).delete({ force: true });
    } catch (_) {
      /* ignore */
    }
  });

  it("exports query results to a sharded CSV under the wildcard uri", async (t) => {
    if (!GCS_HOST_ROOT) {
      t.skip("BQEMU_GCS_HOST_ROOT not set (run via `make test-e2e-nodejs`)");
      return;
    }

    await client.query(
      `CREATE TABLE \`${PROJECT}.${DATASET}.src\` (id INT64, name STRING)`,
    );
    await client.query(
      `INSERT INTO \`${PROJECT}.${DATASET}.src\` (id, name) ` +
        `VALUES (1, 'alpha'), (2, 'beta'), (3, 'gamma')`,
    );

    await client.query(
      `EXPORT DATA OPTIONS (` +
        `uri = 'gs://${BUCKET}/export_nodejs/*.csv', ` +
        `format = 'CSV', overwrite = true) AS ` +
        `SELECT id, name FROM \`${PROJECT}.${DATASET}.src\` ORDER BY id`,
    );

    const shard = path.join(GCS_HOST_ROOT, BUCKET, "export_nodejs", "000000000000.csv");
    assert.ok(fs.existsSync(shard), `expected export shard at ${shard}`);

    const lines = fs
      .readFileSync(shard, "utf8")
      .split("\n")
      .map((l) => l.replace(/\r$/, ""))
      .filter((l) => l.length > 0);
    assert.equal(lines[0], "id,name");
    assert.deepEqual(lines.slice(1), ["1,alpha", "2,beta", "3,gamma"]);
  });
});

File: tests/e2e/go_client/export_data_test.go
// EXPORT DATA → Cloud Storage (CSV) E2E against a live bqemulator container.
//
// EXPORT DATA runs as a query job: the inner SELECT is materialised and
// written to the wildcard `uri` under BQEMU_GCS_LOCAL_ROOT. With a single
// output shard the `*` expands to a 12-digit zero-padded counter, so
// `export_go/*.csv` becomes `export_go/000000000000.csv`. The Makefile
// target (`test-e2e-go`) bind-mounts a host directory onto
// /var/lib/bqemu-gcs and exports its host path as BQEMU_GCS_HOST_ROOT, so
// we read the exported file straight off the mount.

package e2e

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/option"
)

func TestExportDataToCSV(t *testing.T) {
	hostRoot := os.Getenv("BQEMU_GCS_HOST_ROOT")
	if hostRoot == "" {
		t.Skip("BQEMU_GCS_HOST_ROOT not set (run via `make test-e2e-go`)")
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	const project = "e2e-go-export"
	const bucket = "g1-e2e"

	client, err := bigquery.NewClient(
		ctx,
		project,
		option.WithEndpoint(bqAPIBase()),
		option.WithoutAuthentication(),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	defer client.Close()

	ds := client.Dataset("export_go_ds")
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() { _ = ds.DeleteWithContents(ctx) }()

	src := ds.Table("src")
	if err := src.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType, Required: true},
			{Name: "name", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create table: %v", err)
	}

	type row struct {
		ID   int64  `bigquery:"id"`
		Name string `bigquery:"name"`
	}
	if err := src.Inserter().Put(ctx, []row{{1, "alpha"}, {2, "beta"}, {3, "gamma"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}

	exportSQL := "EXPORT DATA OPTIONS (" +
		"uri = 'gs://" + bucket + "/export_go/*.csv', " +
		"format = 'CSV', overwrite = true) AS " +
		"SELECT id, name FROM export_go_ds.src ORDER BY id"
	job, err := client.Query(exportSQL).Run(ctx)
	if err != nil {
		t.Fatalf("export run: %v", err)
	}
	status, err := job.Wait(ctx)
	if err != nil {
		t.Fatalf("export wait: %v", err)
	}
	if err := status.Err(); err != nil {
		t.Fatalf("export job failed: %v", err)
	}

	shard := filepath.Join(hostRoot, bucket, "export_go", "000000000000.csv")
	data, err := os.ReadFile(shard)
	if err != nil {
		t.Fatalf("read shard %s: %v", shard, err)
	}

	var lines []string
	for _, l := range strings.Split(string(data), "\n") {
		l = strings.TrimRight(l, "\r")
		if l != "" {
			lines = append(lines, l)
		}
	}
	want := []string{"id,name", "1,alpha", "2,beta", "3,gamma"}
	if len(lines) != len(want) {
		t.Fatalf("unexpected lines: got %v want %v", lines, want)
	}
	for i := range want {
		if lines[i] != want[i] {
			t.Fatalf("line %d: got %q want %q", i, lines[i], want[i])
		}
	}
}
